File: javascript-solutions/leetcode/p31-nextPermutation.js
/**
 * Leetcode Problem 31 - next permutation
 *
 * Implement next permutation, which rearranges numbers into the lexicographically next greater permutation of numbers.
 * If such an arrangement is not possible, it must rearrange it as the lowest possible order (i.e., sorted in ascending order).
 * The replacement must be in place and use only constant extra memory.
 */




/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var nextPermutation = function(nums) {

  // returns the index of the first value in arr that is not greater than the element after it
  function firstNonAscIndex(arr) {
    if (arr.length < 2) return-1;

    let ci = arr.length - 1;
    let pi = ci - 1;

    do {
      if (arr[pi] < arr[ci]) {
        return pi;
      }
      pi -= 1;
      ci -= 1;
    } while (pi >= 0);
    return -1;
  }

  // find first index of an element in arr that satisfies the predFn, beginning at the end of the array, and ending
  // at startIndex (inclusive)
  function revFindIndex(arr, startIndex, predFn) {
    let i = arr.length - 1;
    while (i >= startIndex) {
      if (predFn(arr[i])) {
        return i;
      }
      i -= 1;
    }
    return -1;
  }

  // swap two elements in the given array at index1 and index2
  function swap(arr, idx1, idx2) {
    const temp = arr[idx1];
    arr[idx1] = arr[idx2];
    arr[idx2] = temp;
  }

  // sort a sub-slice of the given array in place. The sub-slice begins at startIndex to the end of the array.
  function arrSort(arr, startIndex) {
    let slice = arr.slice(startIndex);
    slice = slice.sort((a, b) => a - b);
    for (let i = 0; i < slice.length; i++) {
      arr[startIndex + i] = slice[i];
    }
  }

  // iterate through nums, in reverese order, looking for the first element that is less than the element after it.
  // if such an element is found, call it "fe", then we will use the sub-array to the right of "fe" and swap "fe" with
  // the first element that is greater than "fe", in that sub-array, being sure to iterate the sub-array in REVERSE
  // ORDER. After the swap, we sort the sub array and we are done.
  // If no "fe" was found. than the entire nums array is already in descending order, so sort the whole array and return

  // index of the found element "fe"
  const fIdx = firstNonAscIndex(nums);
  if (fIdx > -1) {
    // search for index of first element > nums[fi]
    const predFn = (num1, num2) => num2 > num1;
    const gtIdx = revFindIndex(nums, fIdx+1, predFn.bind(null, nums[fIdx]));
    if (gtIdx >= 0) {
      swap(nums, fIdx, gtIdx);
      arrSort(nums, fIdx + 1);
    }
  } else {
     arrSort(nums, 0);
  }

};

// const assert = require('assert');
let nums = [1, 2, 3];
nextPermutation(nums);
console.log(nums); // 1, 3 ,2

let nums2 = [3, 2, 1];
nextPermutation(nums2);
console.log(nums2); // 1, 2, 3


let nums3 = [2,2,0,4,3,1];
nextPermutation(nums3);
console.log(nums3);  // 2, 2, 1, 0, 3, 4

let nums4 = [0,0,4,2,1,0];
nextPermutation(nums4);
console.log(nums4);

let nums5 = [4,2,0,2,3,2,0];
nextPermutation(nums5);
console.log(nums5); // [4,2,0,3,0,2,2]

let nums6 = [11,12,0,27,3,11,21,9,0,15,26,27,17,24,0,16,4,17,14,8,15,8,2,16,10,6,6,24,16,2,18,19,6,10,17,10,21,0,11,13,7,7,2,16,24,25,2,20,12,9,20,19];
nextPermutation(nums6);
console.log(nums6);
//exp [11, 12,  0, 27,  3, 11, 21,  9,  0, 15, 26, 27, 17, 24,  0, 16,  4, 17, 14,  8, 15,  8, 2, 16, 10,  6,  6, 24, 16,  2, 18, 19,  6, 10, 17, 10, 21,  0, 11, 13,  7,  7,  2, 16, 24, 25,  2, 20, 12, 19, 9,  20]
//act [11, 12,  0, 27,  3, 11, 21,  9,  0, 15, 26, 27, 17, 24,  0, 16,  4, 17, 14,  8, 15,  8, 2, 16, 10,  6,  6, 24, 16,  2, 18, 19,  6, 10, 17, 10, 21,  0, 11, 13,  7,  7,  2, 16, 24, 25,  2, 20, 12, 19, 20,  9]

File: rust-solutions/leetcode/src/bin/p31-next-permutation.rs
/// # 31 Next Permutation
/// Implement next permutation, which rearranges numbers into the lexicographically next greater permutation of numbers.
/// If such an arrangement is not possible, it must rearrange it as the lowest possible order (i.e., sorted in ascending order).
/// The replacement must be in place and use only constant extra memory.
///
/// ## Example 1
///
/// ```
/// Input: nums = [1,2,3]
/// Output: [1,3,2]
/// ```
///
/// ## Example 2
///
/// ```
/// Input: nums = [3,2,1]
/// Output: [1,2,3]
/// ```
///
/// ## Example 3
///
/// ```
/// Input: nums = [1]
/// Output: [1]
/// ```

pub fn next_permutation(nums: &mut Vec<i32>) {
    // returns the index of the first element in `nums` that is lexicographically < then the element
    // after it, going in reverse order. If no such element is found, then `None` is returned and
    // you can assume the entire slice is in descending order
    fn r_find_first_non_asc_index(nums: &[i32]) -> Option<usize> {
        nums.windows(2).enumerate().rev().find_map(
            |(i, chunk)| {
                if chunk[0] < chunk[1] {
                    Some(i)
                } else {
                    None
                }
            },
        )
    }

    // start at back of nums and find first non ascending index, if any
    if let Some(idx) = r_find_first_non_asc_index(nums) {
        // find first element in nums that is greater than nums[idx] going in reverse order
        if let Some(gt_idx) = nums.iter().rposition(|n| *n > nums[idx]) {
            nums.swap(idx, gt_idx);
            // sort the remaining numbers to the right of idx in ascending order
            nums[idx + 1..].sort_unstable();
        }
    } else {
        nums.sort_unstable();
    }
}

fn main() {}

#[cfg(test)]
mod tests {
    use crate::next_permutation;

    #[test]
    fn example1() {
        let mut nums = vec![1, 2, 3];
        next_permutation(&mut nums);
        assert_eq!(nums, vec![1, 3, 2]);
    }

    #[test]
    fn example2() {
        let mut nums = vec![3, 2, 1];
        next_permutation(&mut nums);
        assert_eq!(nums, vec![1, 2, 3]);
    }

    #[test]
    fn example3() {
        let mut nums = vec![1];
        next_permutation(&mut nums);
        assert_eq!(nums, vec![1]);
    }

    #[test]
    fn example4() {
        let mut nums = vec![1, 2];
        next_permutation(&mut nums);
        assert_eq!(nums, vec![2, 1]);
    }

    #[test]
    fn example4_1() {
        let mut nums = vec![2, 1];
        next_permutation(&mut nums);
        assert_eq!(nums, vec![1, 2]);
    }

    #[test]
    fn example4_2() {
        let mut nums = vec![1, 1];
        next_permutation(&mut nums);
        assert_eq!(nums, vec![1, 1]);
    }

    #[test]
    fn example5() {
        let mut nums = vec![1, 3, 2];
        next_permutation(&mut nums);
        assert_eq!(nums, vec![2, 1, 3]);
    }

    #[test]
    fn example6() {
        let mut nums = vec![2, 3, 1];
        next_permutation(&mut nums);
        assert_eq!(nums, vec![3, 1, 2]);
    }

    #[test]
    fn example7() {
        let mut nums = vec![5, 4, 7, 5, 3, 2];
        next_permutation(&mut nums);
        assert_eq!(nums, vec![5, 5, 2, 3, 4, 7]);
    }

    #[test]
    fn example8() {
        let mut nums = vec![4, 2, 0, 2, 3, 2, 0];
        next_permutation(&mut nums);
        assert_eq!(nums, vec![4, 2, 0, 3, 0, 2, 2]);
    }

    #[test]
    fn example9() {
        let mut nums = vec![2, 2, 0, 4, 3, 1];
        next_permutation(&mut nums);
        assert_eq!(nums, vec![2, 2, 1, 0, 3, 4]);
    }

    #[test]
    fn example10() {
        let mut nums = vec![0, 0, 4, 2, 1, 0];
        next_permutation(&mut nums);
        assert_eq!(nums, vec![0, 1, 0, 0, 2, 4]);
    }
}
